### ISSUE TITLE
Impossible to mount secrets with `userns-remap` enabled

### ISSUE THREAD

**Issue by tyx** [moby/moby#33844]
Hello guys

**Description**

Impossible to mount secrets  if docker engine run with `userns-remap` option enabled.

**Steps to reproduce the issue:**
1.  Enable `--userns-remap=default` on your docker engine
2. `echo "shhht" | docker secret create secret-data -`
3. `docker service  create --name="redis" --secret source=secret-data,target=secret-data redis:alpine`

Already tried to pass `gid` and `uid` with `dockeremap` values and/or `mode` to 0777 to the `--secret`option.

**Describe the results you received:**
We get this kind of message:
```
container_linux.go:262: starting container process caused "process_linux.go:339: container init caused \"rootfs_linux.go:57: mounting \\\"/var/lib/docker/231072.231072/containers/8cd2fe24b473606465afec29a62fc853924590ac3d1c5e340a72b6ba75b6f318/secrets/da22jy6xkdrq3vcxnkjiemmyv\\\" to rootfs \\\"/var/lib/docker/231072.231072/aufs/mnt/b8b7847927a318e3cdddd9c2f6b5febcf2fb9840b0e5baeb01ac7250d5d0e89a\\\" at \\\"/var/lib/docker/231072.231072/aufs/mnt/b8b7847927a318e3cdddd9c2f6b5febcf2fb9840b0e5baeb01ac7250d5d0e89a/run/secrets/website_building_eventstore_password\\\" caused \\\"operation not permitted\\\"\""
```

**Describe the results you expected:**

It should mount secret without problem

**Output of `docker version`:**

```
Client:
 Version:      17.06.0-ce-rc5
 API version:  1.30
 Go version:   go1.8.3
 Git commit:   b7e4173
 Built:        Tue Jun 20 07:13:24 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.06.0-ce-rc5
 API version:  1.30 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   b7e4173
 Built:        Tue Jun 20 07:12:15 2017
 OS/Arch:      linux/amd64
 Experimental: true
```

**Output of `docker info`:**

```
Containers: 5
 Running: 0
 Paused: 0
 Stopped: 5
Images: 15
Server Version: 17.06.0-ce-rc5
Storage Driver: overlay
 Backing Filesystem: extfs
 Supports d_type: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host ipvlan macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: active
 NodeID: s0k8crl0u5mzzlby48ducritq
 Is Manager: true
 ClusterID: yn71r0umggqp1hfdaopewwqab
 Managers: 1
 Nodes: 1
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
  Force Rotate: 0
 Root Rotation In Progress: false
 Node Address: 192.168.99.1
 Manager Addresses:
  192.168.99.1:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: cfb82a876ecc11b5ca0977d1733adbe58599088a
runc version: 2d41c047c83e09a6d61d464906feb2a2f3c52aa4
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
 userns
Kernel Version: 4.4.0-79-generic
Operating System: Ubuntu 16.04.2 LTS
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 7.673GiB
Name: ubuntyx
ID: PEWA:5Q5N:KJYJ:YRUD:MLBW:OR5U:AGFV:CTS4:PCNX:DLD4:AMLX:U45O
Docker Root Dir: /var/lib/docker/165536.165536
Debug Mode (client): false
Debug Mode (server): false
Username: tyxou
Registry: https://index.docker.io/v1/
Experimental: true
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

Thanks you and let me know I can test anything to help

**Comment by thaJeztah** [comment]
Thanks for reporting

ping @diogomonica @cyli @ehazlett PTAL

**Comment by cyli** [comment]
Taking a hint from https://github.com/moby/moby/issues/28859 and https://github.com/moby/moby/issues/25847#issuecomment-240847492, it looks like the problem lies in https://github.com/moby/moby/blob/master/daemon/container_operations_unix.go#L168 - we pass the `nosuid,noexec` options to the tmpfs, and it seems like we need the world executable bit in order to do directory traversals once in `/var/lib/docker/<remappeduid>.<remappedgid>/containers/<containerid>`.

**Comment by justincormack** [comment]
@cyli commented on the PR thats not the right solution, and nothing to do with world execute, it is about executing programs on the `tmpfs`.

You could set world execute, you might have to change the `umask` to do that, but its not ideal I don't think. It would be better to set the ownership to the remapped uid, gid I would think.

**Comment by cyli** [comment]
@justincormack The `tmpfs` ownership is already set to the remapped dockerd root UID and GID, and the secrets within are set to the specified user UID/GID.

**Comment by cyli** [comment]
@justincormack Neither `nodev,nosuid`, and `nodev,noexec`, only `nodev,`.  :|  I have zero clue why.

Leaving either `nosuid` or `noexec` in there still gets me the following in the docker logs:

```
Jun 30 22:29:54 cyli-debug dockerd[12677]: time="2017-06-30T22:29:54.321103674Z" level=error msg="fatal task error" error="starting container failed: oci runtime error: container_linux.go:262: starting container process caused \"process_linux.go:339: container init caused \\\"rootfs_linux.go:57: mounting \\\\\\\"/var/lib/docker/165536.165536/containers/79b005f63ca7ba7c501267ba6055cc7df53bf3a5fb4f7ffe87fe3ae260033f44/secrets/i4jt6olb5fo3c7vnnatvdbptf\\\\\\\" to rootfs \\\\\\\"/var/lib/docker/165536.165536/aufs/mnt/ab247415adf6123876b9a6cb9886ca742d492b563f36cc582249e20cf19f051a\\\\\\\" at \\\\\\\"/var/lib/docker/165536.165536/aufs/mnt/ab247415adf6123876b9a6cb9886ca742d492b563f36cc582249e20cf19f051a/run/secrets/secret-data\\\\\\\" caused \\\\\\\"operation not permitted\\\\\\\"\\\"\"\n" module="node/agent/taskmanager" node.id=fbb4bbkqnze7euj1t0mjil1ga service.id=ox6kgn0cq7ojkgkwsmwyj0yo9 task.id=ji0gnw51xflm4zgr62u2ir5y8
```

As far as I can tell everything seems to be the correct ownership and permissions?  https://asciinema.org/a/k6TRS5r8rRTeR2calh97FXncV

In particular:
```
├── [drwx------ 165536   165536  ]  41dd6b0516b24215c070af561aaff228c2163482ee24807a937c2e7e6eb3e563                                                                                                                                                                                                                          
│   ├── [-rw-r----- root     root    ]  41dd6b0516b24215c070af561aaff228c2163482ee24807a937c2e7e6eb3e563-json.log                                                                                                                                                                                                             
│   ├── [drwx------ 165536   165536  ]  checkpoints                                                                                                                                                                                                                                                                           
│   ├── [-rw-r--r-- root     root    ]  config.v2.json                                                                                                                                                                                                                                                                        
│   ├── [-rw-r--r-- root     root    ]  hostconfig.json                                                                                                                                                                                                                                                                       
│   ├── [-rw-r--r-- 165536   165536  ]  hostname                                                                                                                                                                                                                                                                              
│   ├── [-rw-r--r-- 165536   165536  ]  hosts                                                                                                                                                                                                                                                                                 
│   ├── [-rw-r--r-- 165536   165536  ]  resolv.conf                                                                                                                                                                                                                                                                           
│   ├── [-rw-r--r-- root     root    ]  resolv.conf.hash                                                                                                                                                                                                                                                                      
│   ├── [drwxrwxrwt 165536   165536  ]  secrets                                                                                                                                                                                                                                                                               
│   │   └── [-r--r--r-- 165536   165536  ]  i4jt6olb5fo3c7vnnatvdbptf 
```

**Comment by justincormack** [comment]
Could you run `strace` on that failing `mount`? And a bit before. Actually I think I know what the issue may be, I think `runc` may be buggy about how it does remounts possibly in some circumstances. But its odd it is only with userns. It may be worth trying to replicate it just with `runc`.

**Comment by shouze** [comment]
@justincormack @cyli is it also possible that userns don't remap the root user?

I've seen in some circumstances files that should belong to root:root inside the container belonging to nobody:nogroup  instead whilst in the host uid:gid are the ones of subuid/subgid of dockremap.

**Comment by cyli** [comment]
Update - I still haven't figured it out, but with a lot of help from @crosbymichael, I got this to replicate in runc, both the version that comes with docker 17.06 and in runc master.

Here's a little script I wrote to replicate the issue:

```
#!/usr/bin/env bash

runcCmd="${1}"
testDir="${2}"
remappedUID=1000
remappedGID=1000

# ---- make basic runc container ----
mkdir -p "${2}/rootfs"
cd "${2}"
docker export $(docker create busybox) | tar -C rootfs -xvf -

# make sure everything is a+x all the way up to /, otherwise the rootfs won't even mount
currentDir=$(pwd)
while [ "${currentDir}" != "/" ]; do
    chmod a+x "${currentDir}"
    currentDir=$(dirname ${currentDir})
done
chown -R "${remappedUID}.${remappedGID}" rootfs

"${runcCmd}" spec

# enable userns in the runc spec
sed -i '/"linux":/a "uidMappings":\[\{"hostID": 1000, "containerID":0, "size": 1000\}\], "gidMappings": \[\{"hostID": 1000, "containerID":0, "size": 1000\}\],' config.json
sed -i '/"namespaces":/a \{ "type": "user" \},' config.json

# make a tmpfs
mkdir secretTmpfs
chown -R "${remappedUID}.${remappedGID}" secretTmpfs
chmod 0700 secretTmpfs

mount -t tmpfs -o "rw,nosuid,nodev,noexec,uid=${remappedUID},gid=${remappedGID}" tmpfs secretTmpfs/
echo hello > secretTmpfs/secretfile
chown -R "${remappedUID}.${remappedGID}" secretTmpfs/secretfile
chmod 0444 secretTmpfs/secretfile
mount -t tmpfs -o "remount,ro,nosuid,nodev,noexec,uid=${remappedUID},gid=${remappedGID}" tmpfs secretTmpfs/

# mount the tmpfs in the container
sed -i '/"mounts":/a \{"destination": "/run/secrets/secret-data", "type": "bind", "source": "secretTmpfs/secretfile", "options": \[ "rbind", "ro" \]\},' config.json


# run runc
"${runcCmd}" --debug --log ../output run 0

if [ "${3}" == "clean" ]; then
    umount secretTmpfs
    cd ../
    rm -rf "${testDir}"
fi
```

If I run it:
```
./replicate.sh ~/runc testRunc clean
...
container_linux.go:265: starting container process caused "process_linux.go:348: container init caused \"rootfs_linux.go:57: mounting \\\"/root/testRunc/secretTmpfs/secretfile\\\" to rootfs \\\"/root/testRunc/rootfs\\\" at \\\"/root/testRunc/rootfs/run/secrets/secret-data\\\" caused \\\"operation not permitted\\\"\""
```

With some additional logging in runc, it looks like https://github.com/opencontainers/runc/blob/master/libcontainer/rootfs_linux.go#L779 is the call that fails.

`m` looks like:  ` &{/root/testRunc/secretTmpfs/secretfile /root/testRunc/rootfs/run/secrets/secret-data bind 20481 []   0 [] []}`

I printed out the permissions and uid/gid (from within the ns, anyway) of the source, destination, and rootfs:

```
10 m.Source:
	/root/testRunc/secretTmpfs/secretfile: -r--r--r-- 0.0
	/root/testRunc/secretTmpfs: dtrwxrwxrwx 0.0
	/root/testRunc: drwxr-xr-x 65534.65534
	/root: drwx--x--x 65534.65534

11 m.Destination:
	/root/testRunc/rootfs/run/secrets/secret-data: -r--r--r-- 0.0
	/root/testRunc/rootfs/run/secrets: drwxr-xr-x 0.0
	/root/testRunc/rootfs/run: drwxr-xr-x 0.0
	/root/testRunc/rootfs: drwxr-xr-x 0.0
	/root/testRunc: drwxr-xr-x 65534.65534
	/root: drwx--x--x 65534.65534

12 rootfs:
	/root/testRunc/rootfs: drwxr-xr-x 0.0
	/root/testRunc: drwxr-xr-x 65534.65534
	/root: drwx--x--x 65534.65534
```

Nothing seems to be missing the executable bit.

And doing an `ls` from the host, if I run that script without the `clean` param:
```
root@cyli-debug:~# ls -la /root/testRunc/
total 16
drwxr-xr-x  4 root root 4096 Jul 11 01:08 .
drwx--x--x  9 root root 4096 Jul 11 01:08 ..
-rw-r--r--  1 root root 2883 Jul 11 01:08 config.json
drwxr-xr-x 13 1000 1000 4096 Jul 11 01:08 rootfs
drwxrwxrwt  2 1000 1000   60 Jul 11 01:08 secretTmpfs
root@cyli-debug:~# ls -la /root/testRunc/rootfs/run/secrets/secret-data
-rwxr-xr-x 1 1000 1000 0 Jul 11 01:08 /root/testRunc/rootfs/run/secrets/secret-data
root@cyli-debug:~# ls -la /root/testRunc/secretTmpfs/secretfile
-r--r--r-- 1 1000 1000 6 Jul 11 01:08 /root/testRunc/secretTmpfs/secretfile
```

So not sure what's up.

**Comment by cyli** [comment]
@tycho says that the problem in runc is that once you're an unprivileged user (and I think the mount happens once we've changed namespaces), `atime`, `readonly`, `nodev`, `nosuid`, and `noexec` flags can't be changed via `MS_REMOUNT`.

I set `{"destination": "/run/secrets/secret-data", "type": "bind", "source": "secretTmpfs/secretfile", "options": \[ "rbind", "ro" \]\}` in the script above because from dumping the runc configs that docker creates for these containers, that seems to be what the secret gets mounted as.

It looks like my remount command in the script was typoed:  It looks like it *should* be `mount -t tmpfs -o "remount,ro,uid=${remappedUID},gid=${remappedGID}" tmpfs secretTmpfs/` instead (no `nosuid` or `noexec` options), which is I think what https://github.com/moby/moby/blob/master/daemon/container_operations_unix.go#L219 is trying to do.

Indeed, if I change that replicate script to use `mount -t tmpfs -o "remount,ro,uid=${remappedUID},gid=${remappedGID}" tmpfs secretTmpfs/` when remounting, the runc container starts fine and the secret is readable.

I think the reason this is failing in docker is that the remount isn't actually working.  I dumped the output of the `mount` command immediately after this block:  https://github.com/moby/moby/blob/master/daemon/container_operations_unix.go#L219, and it still says `tmpfs on /var/lib/docker/165536.165536/containers/d517162eff713b6d19777864a9e60b0a089b1d18baa4e30f9343a054d53fa17e/secrets type tmpfs (rw,nosuid,nodev,noexec,relatime,uid=165536,gid=165536)`, instead of being `ro`.

I think that's why my [solution above, with removing `noexec` and `nosuid` for the mounting](https://github.com/moby/moby/issues/33844#issuecomment-312361766), worked, because it never gets remounted.  I'll dig into the remounting code some more.

**Comment by justincormack** [comment]
Ah, thats kind of where I thought the issue might be. But it is a bit odd as the flags are not being changed by `remount` they are actually being kept the same... Will take a look tomorrow.

**Comment by tych0** [comment]
I think the flags are being changed,

https://github.com/moby/moby/blob/master/daemon/container_operations_unix.go#L168

first mounts the tmpfs with nodev,nosuid,noexec, and then the line @cyli linked to later remounts it, but drops those flags, which I think results in the EPERM.

I wonder if this completely untested patch fixes it: https://github.com/tych0/moby/commit/def6ba0ee54fd5017f3734b3a585ef69e981c4c5

**Comment by tych0** [comment]
Oof, sorry, bug in that one, maybe this one? https://github.com/tych0/moby/commit/0de568e8445ead87c75e5be6b9d4d8d22341fefb

**Comment by justincormack** [comment]
Yes, that would be a potential problem; was confused as the original script above keeps the flags, but was then amended.

**Comment by cyli** [comment]
Ok, I've figured out why that [remount line](https://github.com/moby/moby/blob/master/daemon/container_operations_unix.go#L219)  wasn't working:

https://github.com/moby/moby/blob/master/pkg/mount/mounter_linux.go#L30 is the code that eventually gets called, and none of those if blocks are true.  We are trying to remount read-only, but this isn't a bind.  If I change `broflags = unix.MS_BIND | unix.MS_RDONLY` to `broflags = unix.MS_RDONLY`, then the remount happens (if I type `mount` I see `tmpfs on /var/lib/docker/165536.165536/containers/4326ecd16201df3f85d83cc1fbefe67486949b90376ea809674ba44e00d4580b/secrets type tmpfs (ro,relatime,uid=165536,gid=165536)`, and the container successfully starts instead of failing.

I don't know that that's the right fix, though.

**Comment by justincormack** [comment]
@cyli no thats not the fix, the remount-bind-readonly "bro" is a special case see the note under remount in http://man7.org/linux/man-pages/man2/mount.2.html which is why this is there.

**Comment by justincormack** [comment]
I do think there are some missing cases there though...

**Comment by cyli** [comment]
So from a discussion with @tych0 on slack, it looks like what we want to do to fix this is:

1) @tycho's patch to remount with `ro,nosuid,noexec` in https://github.com/moby/moby/blob/master/daemon/container_operations_unix.go#L219.
2) fix https://github.com/moby/moby/blob/master/pkg/mount/mounter_linux.go#L30 so that the above actually works.
    - not sure what additional case to add here to make sure the remounting happens
3) figure out where the bind parameters are passed to runc for secrets, and also add `nosuid,noexec`
    - @tycho traced this to https://github.com/moby/moby/blob/master/container/container_unix.go#L220 - the extra options can probably be passed in the `Data` field
    - that eventually gets translated to the OCI runtime spec here:  https://github.com/moby/moby/blob/master/daemon/oci_linux.go#L483, although it looks like `Data` is only respected and passed on if `Source` is `"tmpfs"`, which in the secrets case it isn't - it's a path, so currently [`["ro", "rbind"]`](https://github.com/moby/moby/blob/master/daemon/oci_linux.go#L553) are the only options that get set.

**Comment by justincormack** [comment]
You should not need any options in the final bind mount, as options in bind
mounts other than recursive are ignored and remain the same.

**Comment by tych0** [comment]
Yes, except that for whatever reason runc also does an MS_REMOUNT of the secrets mount, which fails if you don't pass the additional flags:

```
5174  mount("/home/ubuntu/tester/secretTmpfs/secretfile", "/home/ubuntu/tester/rootfs/run/secrets/secret-data", 0xc820033bc0, MS_RDONLY|MS_BIND|MS_REC, NULL) = 0
5174  mount("/home/ubuntu/tester/secretTmpfs/secretfile", "/home/ubuntu/tester/rootfs/run/secrets/secret-data", 0xc820033bc5, MS_RDONLY|MS_REMOUNT|MS_BIND|MS_REC, NULL) = -1 EPERM (Operation not permitted)
```

**Comment by tych0** [comment]
Although that's probably a bug, and we should just fix that instead of having to pass the flags in 3.

**Comment by justincormack** [comment]
There are a lot of issues with the `mount` code in `runc` sadly... I plan to write proper test cases for it, as currently making changes is very difficult.

**Comment by justincormack** [comment]
I think https://github.com/moby/moby/pull/34077 should fix the not changing the data options in mount. Untested though...

**Comment by cyli** [comment]
@justincormack It does indeed, thank you.  It by itself also fixes the secrets remounting issue, since it's remounting as just `tmpfs on /var/lib/docker/165536.165536/containers/b87f1a3c05ec693c927cbdf1d53c396111c6765e926d0b0a75a258070b26056f/secrets type tmpfs (ro,relatime,uid=165536,gid=165536)`, which matches the parameters in runc (no `nosuid` or `noexec`)

**Comment by justincormack** [comment]
As it is read only, `nosuid` does not really matter, as nothing will be `suid` anyway. Technically if your secret was an executable not having `noexec` would mean you could execute it, but thats not terrible. So I think this alone is probably sufficient for 17.06.1. It would be nice to pass through the right options eventually...

**Comment by shouze** [comment]
I guess 17.06.1 release will close this one, any ETA?

**Comment by justincormack** [comment]
I think within a week.

**Comment by thaJeztah** [comment]
Given that this is merged in master, I'll close this. The patch will also be included in Docker 17.06.1

**Comment by tyx** [comment]
Thanks for this quick fix !
